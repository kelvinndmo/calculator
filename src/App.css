*,
*::before,
*::after {
  box-sizing: border-box;
  font-family: sans-serif;
  font-weight: normal;
}

body {
  display: flex;
  justify-content: center;
  align-items: center;
  /* background: linear-gradient(to right, rgb(15, 15, 15), #17d365); */
}

.calculator-grid {
  position: relative;
  display: grid;
  justify-content: center;
  margin-top: 50px;
  align-content: center;
  grid-template-columns: repeat(4, 100px);
  grid-template-rows: minmax(120px, auto) repeat(5, 100px);
  grid-gap: 0.5rem;
  padding: 0.5rem;
  border: solid 1px #97d8b2;
  border-radius: 0.5rem;
}

.calculator-grid > button {
  cursor: pointer;
  font-size: 2rem;
  border: 1px solid white;
  outline: none;
  background-color: rgba(255, 255, 255, 0.75);
  /* padding: 2rem; */
  margin: 0.3rem;
  border-radius: 0.5rem;
  align-content: center;
  border: solid 1px #97d8b2;
}

.calculator-grid > button:hover {
  background-color: rgba(255, 255, 2555, 0.9);
}
.span-two {
  grid-column: span 2;
}

.output {
  grid-column: 1/-1;
  background-color: rgba(255, 255, 255, 0.75);
  border: solid 1px #97d8b2;
  display: flex;
  align-items: flex-end;
  justify-content: space-around;
  flex-direction: column;
  padding: 10px;
  word-wrap: break-word;
  word-break: break-all;
  border-radius: 0.5rem;
  margin-bottom: -0.3rem;
}

.output .previous-operand {
  /* color: rgba(255, 255, 255, 0.75); */
  font-size: 1.5rem;
}
.output .current-operand {
  /* color: white; */
  font-size: 2.5rem;
}
