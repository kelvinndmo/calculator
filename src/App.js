import React from "react";
import logo from "./logo.svg";
import "./App.css";

class App extends React.Component {
  state = {
    previousText: "",
    currentText: "",
    currentValue: "",
    previousValue: "",
    operation: ""
  };

  appendNumber = e => {
    let number = e.target.value;
    const { currentValue } = this.state;
    if (number === "." && currentValue.includes(".")) return;
    this.setState({
      currentValue: currentValue.toString() + number.toString()
    });
  };

  chooseOperation = e => {
    const { currentValue, previousValue } = this.state;
    let operation = e.target.value;
    if (currentValue === "") return;
    if (previousValue !== "") {
      return this.calculate();
    }

    this.setState({ operation });
    this.setState({ previousValue: currentValue });
    this.setState({ currentValue: "" });
  };

  delete = () => {
    const { currentValue } = this.state;
    const newValue = currentValue.toString().slice(0, -1);
    this.setState({ currentValue: newValue });
  };

  clear = () => {
    this.setState({ currentValue: "" });
    this.setState({ previousValue: "" });
    this.setState({ operation: "" });
  };

  updateDisplay = () => {
    const { previousValue, operation } = this.state;
    if (operation !== null) {
      this.setState({ previousText: previousValue });
    }
  };

  getDisplayNumber = number => {
    const stringNumber = number.toString();
    const intergerDit = parseFloat(stringNumber.split(".")[0]);
    const decimalDigits = stringNumber.split(".")[1];

    let IntergerValue;
    if (isNaN(intergerDit)) {
      IntergerValue = "";
    } else {
      IntergerValue = intergerDit.toLocaleString("en", {
        maximumFractionDigits: 0
      });
    }

    if (decimalDigits != null) {
      return `${IntergerValue}.${decimalDigits}`;
    } else {
      return IntergerValue;
    }
  };

  calculate = () => {
    const { operation, previousValue, currentValue } = this.state;
    let computation;
    const previous = parseFloat(previousValue);
    const current = parseFloat(currentValue);

    if (isNaN(previous) || isNaN(current)) return;
    switch (operation) {
      case "+":
        computation = previous + current;
        break;
      case "-":
        computation = previous - current;
        break;
      case "*":
        computation = previous * current;
        break;
      case "/":
        computation = previous / current;
        break;
      default:
        computation = "";
    }

    this.setState({ currentValue: computation });
    this.setState({ operation: "" });
    this.setState({ previousValue: "" });
  };

  equalsSign = () => {
    this.calculate();
    this.updateDisplay();
  };

  render() {
    const { currentValue, previousValue, operation } = this.state;
    return (
      <div className="calc">
        <div className="calculator-grid">
          <div className="output">
            <div className="previous-operand">{`${this.getDisplayNumber(
              previousValue
            )} ${operation}`}</div>
            <div className="current-operand">{currentValue}</div>
          </div>
          <button onClick={this.clear} className="span-two">
            AC
          </button>
          <button onClick={this.delete}>DEL</button>
          <button onClick={this.chooseOperation} value="/">
            /
          </button>
          <button value={1} onClick={this.appendNumber}>
            1
          </button>
          <button value={2} onClick={this.appendNumber}>
            2
          </button>
          <button value={3} onClick={this.appendNumber}>
            3
          </button>
          <button onClick={this.chooseOperation} value={"*"}>
            *
          </button>
          <button value={4} onClick={this.appendNumber}>
            4
          </button>
          <button value={5} onClick={this.appendNumber}>
            5
          </button>
          <button value={6} onClick={this.appendNumber}>
            6
          </button>
          <button onClick={this.chooseOperation} value={"+"}>
            +
          </button>
          <button value={7} onClick={this.appendNumber}>
            7
          </button>
          <button value={8} onClick={this.appendNumber}>
            8
          </button>
          <button value={9} onClick={this.appendNumber}>
            9
          </button>
          <button onClick={this.chooseOperation} value={"-"}>
            -
          </button>
          <button value={"."} onClick={this.appendNumber}>
            .
          </button>
          <button value={0} onClick={this.appendNumber}>
            0
          </button>
          <button onClick={this.equalsSign} className="span-two">
            =
          </button>
        </div>
      </div>
    );
  }
}

export default App;
