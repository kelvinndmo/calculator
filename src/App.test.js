import React from "react";
import { shallow } from "enzyme";
import App from "./App";

let wrapper;

beforeEach(() => {
  wrapper = shallow(<App />);
  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });
});

test("should render the component App", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should append a number when clicked", () => {
  expect(wrapper.state("currentValue")).toBe("2");
});

test("should append operator when operator is clicked", () => {
  wrapper
    .find("button")
    .at(2)
    .simulate("click", {
      target: { value: "/" }
    });
  expect(wrapper.state("operation")).toBe("/");
});

test("should call calculate when equal sign is presses", () => {
  //we called with wrapper with one in the beforeEach funtion let's call
  // another sign operator then ccall calculate to calculate the result by
  // simulating the equals button
  wrapper
    .find("button")
    .at(2)
    .simulate("click", {
      target: { value: "/" }
    });

  wrapper
    .find("button")
    .at(3)
    .simulate("click", {
      target: { value: "1" }
    });

  //lets now simulate the = button
  wrapper
    .find("button")
    .at(17)
    .simulate("click", {});

  expect(wrapper.state("currentValue")).toBe(2);
});

test("should multiply when a multiplication sign is clicked", () => {
  wrapper
    .find("button")
    .at(6)
    .simulate("click", {
      target: { value: "*" }
    });

  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });

  //lets now simulate the = button
  wrapper
    .find("button")
    .at(17)
    .simulate("click", {});
  //expect 2*2 to be 4
  expect(wrapper.state("currentValue")).toBe(4);
});

test("should add when a addition sign is clicked", () => {
  wrapper
    .find("button")
    .at(10)
    .simulate("click", {
      target: { value: "+" }
    });

  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });

  //lets now simulate the = button
  wrapper
    .find("button")
    .at(17)
    .simulate("click", {});
  //expect 2+2 to be 4
  expect(wrapper.state("currentValue")).toBe(4);
});

test("should substract when a substraction sign is clicked", () => {
  wrapper
    .find("button")
    .at(14)
    .simulate("click", {
      target: { value: "-" }
    });

  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });

  //lets now simulate the = button
  wrapper
    .find("button")
    .at(17)
    .simulate("click", {});
  //expect 2-2 to be 0
  expect(wrapper.state("currentValue")).toBe(0);
});

test("should test that the input field is cleared", () => {
  //simulate the clear button
  wrapper
    .find("button")
    .at(0)
    .simulate("click", {});
  expect(wrapper.state("currentValue")).toBe("");
  expect(wrapper.state("previousValue")).toBe("");
  expect(wrapper.state("operation")).toBe("");
});

test("should delete when delete button is clicked", () => {
  wrapper
    .find("button")
    .at(1)
    .simulate("click", {});
  expect(wrapper.state("currentValue")).toBe("");
  expect(wrapper.state("previousValue")).toBe("");
  expect(wrapper.state("operation")).toBe("");
});

test("should automatically calculate when more than one operation is passed", () => {
  wrapper
    .find("button")
    .at(14)
    .simulate("click", {
      target: { value: "-" }
    });

  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });
  wrapper
    .find("button")
    .at(14)
    .simulate("click", {
      target: { value: "-" }
    });
  wrapper
    .find("button")
    .at(14)
    .simulate("click", {
      target: { value: "-" }
    });
  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });
  expect(wrapper.state("currentValue")).toBe("2");
});

test("should previous text should be empty when an operator has not been input", () => {
  expect(wrapper.state("previousText")).toBe("");
});

test("should add decimals when a addition sign is clicked with dot character", () => {
  wrapper
    .find("button")
    .at(10)
    .simulate("click", {
      target: { value: "+" }
    });

  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });
  wrapper
    .find("button")
    .at(15)
    .simulate("click", {
      target: { value: "." }
    });
  wrapper
    .find("button")
    .at(4)
    .simulate("click", {
      target: { value: "2" }
    });

  //lets now simulate the = button
  wrapper
    .find("button")
    .at(17)
    .simulate("click", {});
  //expect 2+2 to be 4
  expect(wrapper.state("currentValue")).toBe(4.2);
});

test("should return when switch has no operation", () => {
  wrapper
    .find("button")
    .at(17)
    .simulate("click", {});
  expect(wrapper.state("operation")).toBe("");
});
