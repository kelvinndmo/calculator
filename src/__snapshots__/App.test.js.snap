// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component App 1`] = `
<div
  className="calc"
>
  <div
    className="calculator-grid"
  >
    <div
      className="output"
    >
      <div
        className="previous-operand"
      >
         
      </div>
      <div
        className="current-operand"
      >
        2
      </div>
    </div>
    <button
      className="span-two"
      onClick={[Function]}
    >
      AC
    </button>
    <button
      onClick={[Function]}
    >
      DEL
    </button>
    <button
      onClick={[Function]}
      value="/"
    >
      /
    </button>
    <button
      onClick={[Function]}
      value={1}
    >
      1
    </button>
    <button
      onClick={[Function]}
      value={2}
    >
      2
    </button>
    <button
      onClick={[Function]}
      value={3}
    >
      3
    </button>
    <button
      onClick={[Function]}
      value="*"
    >
      *
    </button>
    <button
      onClick={[Function]}
      value={4}
    >
      4
    </button>
    <button
      onClick={[Function]}
      value={5}
    >
      5
    </button>
    <button
      onClick={[Function]}
      value={6}
    >
      6
    </button>
    <button
      onClick={[Function]}
      value="+"
    >
      +
    </button>
    <button
      onClick={[Function]}
      value={7}
    >
      7
    </button>
    <button
      onClick={[Function]}
      value={8}
    >
      8
    </button>
    <button
      onClick={[Function]}
      value={9}
    >
      9
    </button>
    <button
      onClick={[Function]}
      value="-"
    >
      -
    </button>
    <button
      onClick={[Function]}
      value="."
    >
      .
    </button>
    <button
      onClick={[Function]}
      value={0}
    >
      0
    </button>
    <button
      className="span-two"
      onClick={[Function]}
    >
      =
    </button>
  </div>
</div>
`;
